packages/core/versioning/core-versioned-http-server-internal/src/internal_versioned_route.ts

```diff
@@ -0,0 +1,164 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type {
+  RequestHandler,
+  IRouter,
+  RequestHandlerContextBase,
+  KibanaRequest,
+  KibanaResponseFactory,
+} from '@kbn/core-http-server';
+import type {
+  Version,
+  AddVersionOpts,
+  VersionedRoute,
+  VersionedRouteConfig,
+} from '@kbn/core-versioned-http-server';
+import type { CoreKibanaRequest } from '@kbn/core-http-router-server-internal';
+import type { Method } from './types';
+
+import { validate } from './validate';
+
+type Options = AddVersionOpts<unknown, unknown, unknown, unknown>;
+
+export const VERSION_HEADER = 'TBD';
+
+// This validation is a pass-through so that we can apply our version-specific validation later
+const passThroughValidation = { body: schema.any(), params: schema.any(), query: schema.any() };
+
+export class InternalVersionedRoute implements VersionedRoute {
+  private readonly handlers = new Map<
+    Version,
+    {
+      fn: RequestHandler;
+      options: Options;
+    }
+  >();
+
+  public static from({
+    router,
+    method,
+    path,
+    options,
+  }: {
+    router: IRouter;
+    method: Method;
+    path: string;
+    options: VersionedRouteConfig<Method>;
+  }) {
+    return new InternalVersionedRoute(router, method, path, options);
+  }
+
+  private constructor(
+    private readonly router: IRouter,
+    public readonly method: Method,
+    public readonly path: string,
+    public readonly options: VersionedRouteConfig<Method>,
+    // TODO: Make "true" dev-only
+    private readonly validateResponses: boolean = true
+  ) {}
+
+  /** This is where we must implement the versioned spec once it is available */
+  private requestHandler = async (
+    ctx: RequestHandlerContextBase,
+    req: KibanaRequest,
+    res: KibanaResponseFactory
+  ) => {
+    const version = req.headers[VERSION_HEADER] as undefined | Version;
+    if (!version) {
+      return res.custom({
+        statusCode: 406,
+        body: `Version expected at [${this.method}] [${this.path}]. Please specify a version in the ${VERSION_HEADER} header.`,
+      });
+    }
+
+    const handler = this.handlers.get(version);
+    if (!handler) {
+      return res.custom({
+        statusCode: 406,
+        body: `No ${version} available for ${this.method} ${this.path}. Available versions are: ${[
+          this.handlers.keys(),
+        ].join(',')}`,
+      });
+    }
+
+    const coreKibanaRequest = req as CoreKibanaRequest;
+    if (handler.options.validate && handler.options.validate.request) {
+      try {
+        const { body, params, query } = validate(
+          req,
+          handler.options.validate.request,
```

Comment:
So, my opinion regarding [your thoughts](https://github.com/elastic/kibana/pull/153543/files#r1146180744):

First, I have to admit I missed that too when we were discussing about this. Kibana requests are instantiated directly with their validation schemas, which is a problem for us here...

Now,

I think the correct approach would be to have our own implementation of `KibanaRequest` that would wrap this `CoreKibanaRequest`. It would delegates (or 'proxy') most of the prop accesses to the underlying `CoreKibanaRequest`, except for `params`, `query` and `body`. That would be the correct way to address the problem here.

That being said, and even if what you did is more of a workaround/hack, I'd say it would be fine-ish if we want to keep it that way.

Reply:
Yeah, I was thinking about using a JS `Proxy` for this purpose. But I know it introduces a performance overhead for access. Manually wrapping felt like overhead.

I'm happy to go the Proxy route if we feel performance does not matter (I think it is about 1/3 the speed of normal access). I concluded that it is better to leave our performance budget as intact as possible.

Reply:
I wouldn't use a proxy, just a boring, manually maintained wrapper with `get`-based accessors.

But, again, just allowing to override the properties as you did is perfectly fine unless we find a reason to want more.